avc/src/planner/planner.cpp

```diff
@@ -1,129 +1,199 @@
-		p.pose.position.y = step.y;
-		path.poses.push_back(p);
-	}
-	path.header.frame_id = "map";
-	//ROS_INFO_STREAM(path.poses.size());
-	path_pub.publish(path);
+    pcl::PCLPointCloud2 pcl_pc2;
+    pcl_conversions::toPCL(*map, pcl_pc2);
+    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::fromPCLPointCloud2(pcl_pc2, *cloud);
+    if(cloud->empty()) {
+        ROS_WARN("Point cloud is empty");
+        rr_platform::speedPtr speedMSG(new rr_platform::speed);
+        rr_platform::steeringPtr steerMSG(new rr_platform::steering);
+        speedMSG->speed = 0.5; //proceed with caution; E-Kill if necessary
+        steerMSG->angle = 0;
+        speed_pub.publish(speedMSG);
+        steer_pub.publish(steerMSG);
+        return;
+    }
+
+    pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+    kdtree.setInputCloud(cloud);
+
+    //build paths and evaluate their weights
+    double weightTotal, cost, angle;
+    int nPathPoses;
+    weightTotal = nPathPoses = 0;
+    vector<double> weightedSteeringRaw;
+    for(int i = 0; i < PATH_ITERATIONS; i++) {
+        vector<double> steerPath;
+        for(int s = 0; s < PATH_STAGES; s++) {
+            angle = steeringSample();
+            for(double t = 0; t <= PATH_STAGE_TIME; t += TIME_INCREMENT) {
+                steerPath.push_back(angle);
+            }
+        }
+
+        sim_path sp = calculatePath(steerPath);
+        cost = calculatePathCost(sp, kdtree);
+
+        weightTotal += 1.0 / cost;
+        nPathPoses++;
+
+        for(int j = 0; j < sp.angles.size(); j++) {
+            if(j < weightedSteeringRaw.size()) {
+                weightedSteeringRaw[j] += sp.angles[j] / cost;
+            } else {
+                weightedSteeringRaw.push_back(sp.angles[j] / cost);
+            }
+        }
+    }
+
+    vector<double> bestSteering;
+    for(int i = 0; i < weightedSteeringRaw.size(); i++) {
+        bestSteering.push_back(weightedSteeringRaw[i] / weightTotal);
+    }
+
+    sim_path bestPath = calculatePath(bestSteering);
+
+    desired_steer_angle = bestPath.angles[0];
+    desired_velocity = steeringToSpeed(desired_steer_angle);
+
+    rr_platform::speedPtr speedMSG(new rr_platform::speed);
+    rr_platform::steeringPtr steerMSG(new rr_platform::steering);
+    speedMSG->speed = desired_velocity;
+    steerMSG->angle = desired_steer_angle;
+    speed_pub.publish(speedMSG);
+    steer_pub.publish(steerMSG);
+
+    nav_msgs::Path pathMsg;
+    for(int i = 0; i < bestPath.poses.size(); i++) {
+        geometry_msgs::PoseStamped p;
+        p.pose.position.x = bestPath.poses[i].x;
+        p.pose.position.y = bestPath.poses[i].y;
+        pathMsg.poses.push_back(p);
+    }
+    pathMsg.header.frame_id = "map";
+    path_pub.publish(pathMsg);
+}
+
+//a convenient scope
+void planner::spin() {
+    normal_distribution<double> nd(0, STEER_STDDEV);
+    steering_gaussian_ptr = &nd;
+    default_random_engine dre;
+    rand_gen_ptr = &dre;
```

Comment:
*warning:* You have entered the realm of random number generation, so I will have a lot of comments here.

1. Why do these need to be pointers?
2. Why can't they just be initialized in the constructor after you've read the *STEER_STDDEV* parameter?
3. std::default_random_engine is implementation defined, which is sketchy as it could change planner performance based on the compiler we use. *mt19937* and *minstd_rand0* are both popular generators.
4. Standard generators take a seed value to their constructor. By not providing a seed value, you are forcing the generator to use its default seed. This actually means you're using the same seed every time the code is run, and you're numbers aren't random at all, but just a very convoluted way to get the same sequence of numbers every time you run. A better (not perfect, but good enough) solution is to pass `std::random_device{}()` to the generators constructor. We can talk about all the details going on here at the next meeting.